lib/src/parallel_prefix_operations.dart

```diff
@@ -0,0 +1,250 @@
+      _oseq[i] <= el;
+    });
+  }
+}
+
+/// BrentKung shaped pp tree
+class BrentKung extends ParallelPrefix {
+  /// BrentKung constructor
+  BrentKung(List<Logic> inps, Logic Function(Logic, Logic) op)
+      : super(inps, 'brent_kung') {
+    final iseq = <Logic>[];
+
+    inps.forEachIndexed((i, el) {
+      iseq.add(addInput('i$i', el, width: el.width));
+      _oseq.add(addOutput('o$i', width: el.width));
+    });
+
+    // Reduce phase
+    var skip = 2;
+    while (skip <= inps.length) {
+      for (var i = skip - 1; i < inps.length; i += skip) {
+        iseq[i] = op(iseq[i - skip ~/ 2], iseq[i]);
+      }
+      skip *= 2;
+    }
+
+    // Prefix Phase
+    skip = largestPow2LessThan(inps.length);
+    while (skip > 2) {
+      for (var i = 3 * (skip ~/ 2) - 1; i < inps.length; i += skip) {
+        iseq[i] = op(iseq[i - skip ~/ 2], iseq[i]);
+      }
+      skip ~/= 2;
+    }
+
+    // Final row
+    for (var i = 2; i < inps.length; i += 2) {
+      iseq[i] = op(iseq[i - 1], iseq[i]);
+    }
+
+    iseq.forEachIndexed((i, el) {
+      _oseq[i] <= el;
+    });
+  }
+}
+
+/// Or scan based on pp tree
+class PPOrScan extends Module {
+  /// Output sequence value
+  Logic get out => output('out');
+
+  /// OrScan constructor
+  PPOrScan(
+      Logic inp,
+      ParallelPrefix Function(List<Logic>, Logic Function(Logic, Logic))
+          ppGen) {
+    inp = addInput('inp', inp, width: inp.width);
+    final u =
+        ppGen(List<Logic>.generate(inp.width, (i) => inp[i]), (a, b) => a | b);
+    addOutput('out', width: inp.width) <= u.val.rswizzle();
+  }
+}
+
+/// Priority Encoder based on pp tree
+class PriorityEncoder extends Module {
+  /// Output sequence value
+  Logic get out => output('out');
+
+  /// PriorityEncoder constructor
+  PriorityEncoder(
+      Logic inp,
+      ParallelPrefix Function(List<Logic>, Logic Function(Logic, Logic))
+          ppGen) {
+    inp = addInput('inp', inp, width: inp.width);
+    final u = PPOrScan(inp, ppGen);
+    addOutput('out', width: inp.width) <= (u.out & ~(u.out << Const(1)));
+  }
+}
+
+/// Adder based on pp tree
+class PPAdder extends Module {
+  /// Output sequence value
+  Logic get out => output('out');
+
+  /// Adder constructor
+  PPAdder(
+      Logic a,
+      Logic b,
+      ParallelPrefix Function(List<Logic>, Logic Function(Logic, Logic))
+          ppGen) {
+    a = addInput('a', a, width: a.width);
+    b = addInput('b', b, width: b.width);
+    final u = ppGen(
+        //                                    generate,    propagate or generate
+        List<Logic>.generate(
+            a.width, (i) => [a[i] & b[i], a[i] | b[i]].swizzle()),
+        (lhs, rhs) => [rhs[1] | rhs[0] & lhs[1], rhs[0] & lhs[0]].swizzle());
+    addOutput('out', width: a.width) <=
+        List<Logic>.generate(a.width,
+                (i) => (i == 0) ? a[i] ^ b[i] : a[i] ^ b[i] ^ u.val[i - 1][1])
+            .rswizzle();
+  }
+}
+
+/// Incrementer based on pp tree
```

Comment:
Can you elaborate more on the incrementer and decrementer, what the input should be and what operation is performed?  Comments like "output sequence" are a little too generic

Reply:
Yes, will do.